### PR TITLE
Add workaroud for 5109

### PR DESCRIPTION
**Pull request by w13915984028**
The issue https://github.com/harvester/harvester/issues/5109 may have no chance to be fixed in Harvester v1.3.0, we add a workaroud to the document.

note: the Workaround part is to be confirmed with LH experts.

### CODE REVIEW

File: docs/troubleshooting/vm.md [hunk 1]
```diff
@@ -243,3 +243,111 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+brw-rw---- 1 root root 8, 16 Feb 21 10:53 4883af80-c202-4529-a2c6-4e7f15fe5a9b
+```
+
+#### Root Cause
+
+The kubelet calls `NodePublishVolume` for the new pod after node/cluster rebooting without first calling `NodeStageVolume`, and the Longhorn CSI plugin bind mounts the regular file at staging target path (previously used by the deleted pod) to target path and returns success.
+
+### Workaround
+
+To address the issue, you can force delete the pod using the command `kubectl delete pod virt-launcher-*** -n namespace --force`.
```
Comment (ejweber) [inline docs/troubleshooting/vm.md:349]: In my testing with a Deployment that is only allowed to run on one node, force deleting the Pod does NOT mitigate the problem. The Deployment controller reschedules the Pod to the same node and `NodePublishVolume` is again called without `NodeStageVolume`.

In addition, even though we no longer see the Pod with kubectl, kubelet can never clean up its PersistentVolume, so Longhorn can never be asked to detach it. This can lead to many additional problems down the line.

After much experimentation, see https://github.com/longhorn/longhorn/issues/8009#issuecomment-1979525625 for the workarounds that helped in the pure Longhorn (with Deployments) reproduce. I think they will be applicable in the Harvester situation as well.

File: docs/troubleshooting/vm.md [hunk 2]
```diff
@@ -243,3 +243,111 @@ After the pod is successfully deleted, the `Start` button becomes visible again
 ### Related Issue
 
 https://github.com/harvester/harvester/issues/4659
+
+## VM Stucks in Starting with Error Messsage `not a device node`
+
+_Available as of v1.3.0_
```
Comment (ibrokethecloud) [inline docs/troubleshooting/vm.md:249]: may be we should reword this to: "Impacted Versions: v1.3.0"?

File: docs/troubleshooting/vm.md (hunk 1 above)
Comment (ibrokethecloud) [reply docs/troubleshooting/vm.md:349]: I can confirm force delete does not solve the issue. I had to restart the kubelet and then force-delete the pod to get reconcile to be retriggered.

Comment (w13915984028) [reply docs/troubleshooting/vm.md]: Sorry, I forgot to remove the workaround, now updated as `TBD`; discussed with @ejweber and are working on a solution for Harvester.

Comment (w13915984028) [reply docs/troubleshooting/vm.md]: @ejweber I copied and updated your option 2 to this PR to as a workaround for Harvester. And add you as the co-author . Thanks.

File: docs/troubleshooting/vm.md [hunk 3]
```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+    pod "virt-launcher-vm1-nxfm4" force deleted
+    ```
+
+    The VM is off now.
+
+    ![](/img/v1.3/troubleshooting/vm-is-off.png)
+
+Node level operation, node by node:
+
+1. [Cordon](../host/host.md#cordoning-a-node) a node.
```
Comment (ejweber) [inline docs/troubleshooting/vm.md:381]: I cordon the node in my workaround because the Deployment I am using will immediately reschedule a pod to the node if I delete one. This prevents the CSI flow from ever getting unstuck.

If you forcefully delete the pods above, will they be potentially rescheduled to the node? If so, should we cordon first?

If there is no danger that the pods will be rescheduled to the node, is cordoning necessary?

File: docs/troubleshooting/vm.md [hunk 4]
```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+    $ kubectl get pvc -A
+    NAMESPACE                  NAME                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS           AGE
+    default                    vm1-disk-0-9gc6h           Bound    pvc-f1798969-5b72-4d76-9f0e-64854af7b59c   1Gi        RWX            longhorn-image-fxsqr   7d22h
+    ```
+
+1. [Stop](#vm-general-operations) the affected VMs from Harvester UI.
+
+   The VM may stuck in `Stopping`, continue the next step.
+
+1. Delete the backing pods forcely.
```
Comment (ejweber) [inline docs/troubleshooting/vm.md:367]: NIT: There is no reason from the Longhorn side that the pods need to be deleted forcefully. Assuming the unmount step proceeds well, the pod that is stuck in `Terminating` will eventually be successfully cleaned up.

**Review by ejweber — COMMENTED** [review]
Overall the workaround looks correct. I have a couple of questions or suggestions, but I don't feel particularly strongly about them, so please just take a look and respond however you think makes sense.

File: docs/troubleshooting/vm.md (hunk 3 above)
Comment (w13915984028) [reply docs/troubleshooting/vm.md:381]: The idea is to stop & clean the affected VM & POD at first, avoid they are scheduled/redeployed to other nodes accidently. 

Then cordon the node to unmount the volumes.

The `cordon` maybe optional if all affected VMs have been stopped&cleaned.

Comment (w13915984028) [reply docs/troubleshooting/vm.md:381]: I need to further consider the scenario or Rancher managed downstream clusters, if a VM is stopped from Harvester UI, will this VM be re-enabled or recreated automatically? That may lead to kind of race.

Comment (ejweber) [reply docs/troubleshooting/vm.md:381]: Ok. I certainly have no concerns leaving the cordon in if we are unsure. I think I mostly wanted to be sure that the VM couldn't be rescheduled to the node BEFORE the cordon. If Harvester will not do that, all good.

File: docs/troubleshooting/vm.md (hunk 4 above)
Comment (ejweber) [reply docs/troubleshooting/vm.md:367]: I expect it might take up to five minutes, since kubelet is in exponential backoff trying to clean up the volumes. In my testing, the next time kubelet tried to unmap the volume after doing the unmount step, it was able to proceed and succeed.

File: docs/troubleshooting/vm.md [hunk 5]
```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+$ ls  /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-732f8496-103b-4a08-83af-8325e1c314b7/ -alth
+total 8.0K
+drwxr-x--- 2 root root  4.0K Feb 21 10:53 .
+drwxr-x--- 4 root root  4.0K Feb 21 10:53 ..
+brw-rw---- 1 root root 8, 16 Feb 21 10:53 4883af80-c202-4529-a2c6-4e7f15fe5a9b
+```
+
+#### Root Cause
+
+The kubelet calls `NodePublishVolume` for the new pod after node/cluster rebooting without first calling `NodeStageVolume`, and the Longhorn CSI plugin bind mounts the regular file at staging target path (previously used by the deleted pod) to target path and returns success.
```
Comment (jillian-maroket) [inline docs/troubleshooting/vm.md:345]:
```suggestion
After the cluster or specific nodes are rebooted, the kubelet calls `NodePublishVolume` for the new pod without first calling `NodeStageVolume`. Moreover, the Longhorn CSI plugin bind mounts the regular file at the staging target path (previously used by the deleted pod) to the target path, and the operation is considered successful.
```

**Review by jillian-maroket — COMMENTED** [review]
Initial review done. I did not touch the "Workaround" content because there are unresolved discussions with the Longhorn team. I'll resume the review once the workaround procedure is finalized.

File: docs/troubleshooting/vm.md [hunk 6]
```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+The keywords `failed to generate spec: not a device node` are found in `containerd` log:
+
+```
+file path: /var/lib/rancher/rke2/agent/containerd/containerd.log
+
+time="2024-02-21T11:24:00.140298800Z" level=error msg="CreateContainer within sandbox \"850958f388e63f14a683380b3c52e57db35f21c059c0d93666f4fdaafe337e56\" for &ContainerMetadata{Name:compute,Attempt:0,} failed" error="failed to generate container \"5ddad240be2731d5ea5210565729cca20e20694e364e72ba14b58127e231bc79\" spec: failed to generate spec: not a device node"
+
+```
+
+After adding debug information to `contaierd`, it shows that error message `not a device node` is upon the file `pvc-3c1b28fb-`.
```
Comment (bk201) [inline docs/troubleshooting/vm.md:311]:
```suggestion
After adding debug information to `containerd`, it shows that error message `not a device node` is upon the file `pvc-3c1b28fb-`.
```

**Review by bk201 — COMMENTED** [review]
overall look good, thanks!

File: docs/troubleshooting/vm.md (hunk 4 above)
Comment (w13915984028) [reply docs/troubleshooting/vm.md:367]: @ejweber 
Reproduced and retested in Harvester, the VM's POD is still stucking on `Terminating` more than 30 minutes later,  after `Stop` the VM on UI. A force deleting of the POD seems better.

```
$ while true; do kubectl get pods; date; sleep 10; done
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
Mon Mar 11 14:44:28 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
Mon Mar 11 14:44:38 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
...
...
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          53m
Mon Mar 11 15:14:45 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          53m
```

```
E0311 15:14:54.856070    2286 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/driver.longhorn.io^pvc-f1798969-5b72-4d76-9f0e-64854af7b59c podName:a676483b-ef5a-4eb4-969b-24038ce26448 nodeName:}" failed. No retries permitted until 2024-03-11 15:16:56.856056804 +0000 UTC m=+3511.059632246 (durationBeforeRetry 2m2s). Error: UnmapVolume.UnmapBlockVolume failed for volume "disk-0" (UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-f1798969-5b72-4d76-9f0e-64854af7b59c") pod "a676483b-ef5a-4eb4-969b-24038ce26448" (UID: "a676483b-ef5a-4eb4-969b-24038ce26448") : blkUtil.DetachFileDevice failed. globalUnmapPath:/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev, podUID: a676483b-ef5a-4eb4-969b-24038ce26448, bindMount: true: failed to remove file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev/a676483b-ef5a-4eb4-969b-24038ce26448: remove /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev/a676483b-ef5a-4eb4-969b-24038ce26448: device or resource busy
```

**Comment by bk201** [comment]
@w13915984028 Can you help fix the suggested changes so we can proceed with the PR? Thanks.

File: docs/troubleshooting/vm.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (w13915984028) [reply docs/troubleshooting/vm.md]: Partially updated.

File: docs/troubleshooting/vm.md [hunk 7]
```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
...
+
+```
+time="2024-02-22T15:15:08.557487376Z" level=error msg="CreateContainer within sandbox \"d23af3219cb27228623cf8168ec27e64e836ed44f2b2f9cf784f0529a7f92e1e\" for &ContainerMetadata{Name:compute,Attempt:0,} failed" error="failed to generate container \"e4ed94fb5e9145e8716bcb87aae448300799f345197d52a617918d634d9ca3e1\" spec: failed to generate spec: get device path: /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19 containerPath: /dev/disk-0 error: not a device node"
+```
+
+This is a CSI related file is an empty file instead of the expected device file. Then the containerd denied the `CreateContainer` request.
```
[the remark this answers is not in the document]
Comment (w13915984028) [reply docs/troubleshooting/vm.md]: Rephased the description. The `Also` is not adpoted, still use `Then`, it is a result of the file error.

**Review by ibrokethecloud — APPROVED** [review]
lgtm. thanks.

**Review by bk201 — APPROVED** [review]

**Review by jillian-maroket — APPROVED** [review]
Approving only for GitHub workflow purposes. The wording in the "Workaround" section can be improved in another PR.

**Comment by bk201** [comment]
Thanks @jillian-maroket, I'll merge this one to publish the doc for the release because it's a known issue with higher severity.
